### DISCUSSION TITLE
Re-send discovery messages on network change

### DISCUSSION THREAD

**Discussion by AlfaBravoX** [arendst/Tasmota#17363] [Ideas]
### PROBLEM DESCRIPTION
This is enhancement request to re-send discovery messages after IP address is changed. When different IP is obtained from DHCP server, Tasmota still publishes days old retained message with invalid IP address, that has been received after boot. For implementations that rely on correct IP address obtained from discovery retained messaged, it is causing applications to missbehave. It would be worth, if discovery messages are re-send on network config change so IP address published in discovery messages always remains correct. 

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): _____
- [X] Tasmota binary firmware version number used: _____
  - [X] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: esptool.py
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
{"NAME":"HW-655 PZEM","GPIO":[0,1472,256,1536,0,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1}
{"Module":{"0":"HW-655 PZEM"}}
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:

```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
15:02:09.576 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS = {"Status":{"Module":0,"DeviceName":"PZEM NH10 ASOL","FriendlyName":["Tasmota"],"Topic":"PZEM-NH10-ASOL","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
15:02:09.583 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS1 = {"StatusPRM":{"Baudrate":9600,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/tasmota-sensors.bin.gz","RestartReason":"Power On","Uptime":"4T04:35:03","StartupUTC":"2022-12-08T09:27:06","Sleep":50,"CfgHolder":4617,"BootCount":72,"BCResetTime":"2022-08-01T19:59:16","SaveCount":1918,"SaveAddress":"F7000"}}
15:02:09.589 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS2 = {"StatusFWR":{"Version":"12.0.2.4(sensors)","BuildDateTime":"2022-07-29T18:04:46","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"381/699"}}
15:02:09.596 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["SSIDw",""],"TelePeriod":10,"Resolution":"558380C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000280","00006000","00004000","00000000"]}}
15:02:09.612 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS4 = {"StatusMEM":{"ProgramSize":678,"Free":324,"Heap":19,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8F9A8787","0415A005","B7F7BFCF","05DA9BC4","64367CC7","00084052","20000000","54000020","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,14,16,17,20,21,24,29,34,62","Sensors":"1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,28,31,34,37,39,40,42,43,45,51,52,55,56,58,59,62,64,66,67,74"}}
15:02:09.620 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS5 = {"StatusNET":{"Hostname":"PZEM-NH10-ASOL-1778","IPAddress":"192.168.4.32","Gateway":"192.168.4.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.4.1","DNSServer2":"0.0.0.0","Mac":"48:55:19:12:86:F2","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
15:02:09.626 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS6 = {"StatusMQT":{"MqttHost":"mysite.com","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_1286F2","MqttUser":"zizkovpi","MqttCount":2,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
15:02:09.635 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS7 = {"StatusTIM":{"UTC":"2022-12-12T14:02:09","Local":"2022-12-12T15:02:09","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":99,"Sunrise":"08:34","Sunset":"16:52"}}
15:02:09.641 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
15:02:09.654 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS10 = {"StatusSNS":{"Time":"2022-12-12T15:02:09","ENERGY":{"TotalStartTime":"2022-08-01T19:59:16","Total":94.662,"Yesterday":0.285,"Today":0.122,"Period": 0,"Power":10,"ApparentPower":15,"ReactivePower":12,"Factor":0.64,"Frequency":50,"Voltage":233.7,"Current":0.066}}}
15:02:09.662 MQT: home-assistant/stat/PZEM-NH10-ASOL/STATUS11 = {"StatusSTS":{"Time":"2022-12-12T15:02:09","Uptime":"4T04:35:03","UptimeSec":362103,"Heap":19,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"POWER":"ON","Wifi":{"AP":1,"SSId":"Na-Hlidce-891w","BSSId":"F4:4E:05:19:A0:C0","Channel":4,"Mode":"11n","RSSI":26,"Signal":-87,"LinkCount":2,"Downtime":"0T00:05:55"}}}
```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Setup 8266 module with DHCP client.

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
force DHCP server to assign different IP address to tasmota DHCP client.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
Retained mqtt messages are stored in mqtt broker. As long Tasmota is online the same IP address is assigned from DHCP and will be used. After a restart it may happen that a different IP address is assigned from DHCP server. Tasmota will use. Tasmota does not store anywhere the assigned IP address from DHCP server, so it is not possible that an old outdated is send from Tasmota.

**Reply by AlfaBravoX** [reply]
Yes, all what your wrote is clear and we would not need to store assigned IP address in Tasmota. This feature request is about to send again discovery message on any `DHCP IP assigned event`. Thats it.

**Comment by Jason2866** [comment]
Why and how should this be done? If you are interested (any time) just query. There is no `DHCP IP assigned event` happening as long Tasmota is running. It happens one time when Tasmota is starting.

**Reply by AlfaBravoX** [reply]
I can live with that and can use STATUS5 query, however integrations that use IP address from discovery may get broken after certain time when DHCP renews. if we cannot track DHCP events then it is difficult to implement, no doubt.

**Comment by Jason2866** [comment]
Tasmota is designed to use mqtt. The `m` in Tasmota stands for mqtt.
People not using mqtt have to live with Tasmota not being optimized for other usage.

**Comment by barbudor** [comment]
@AlfaBravoX 
I don't get your use case
When would tasmota get a new IP address without being disconnected from the network?

If wifi gets disconnected, mqtt is disconnected
On mqtt reconnection, tasmota is supposed to resend the discovery message

So I can't find a hole when it would be needed to send fresh discovery message
But may be I'm missing something

**Comment by AlfaBravoX** [comment]
I was able to simulate by:

1. clear retained discovery message `mosquitto_pub -h localhost -t "tasmota/discovery/{{my_tasmota_ids}}/config" -r -n`
2. after `clear ip dhcp binding *` command issued on Cisco DHCP server (clear all DHCP associations)
3. simulate multiple times (10x) weak WIFI signal with connect and disconnect for nodemcu 8266. 

After this 8266 got new IP address, but no new discovery messages was issued. 
Conclusion: there was for sure severe multiple network interrupts and no new discovery messages.

**Reply by Jason2866** [reply]
Corner case, why should DHCP server loose the associations and at the same time Tasmota device does a reconnect?
My answer for this fix the bad network ;-)
And no this is not only blabla talking. Since Tasmota monitors wifi and mqtt reconnects, i can say that it is normal for Tasmota that there is only about 2 or 3 reconnects a year! for wifi with a decent wifi network.

**Reply by Jason2866** [reply]
Repeating Tasmota is made for using mqtt. If you need the IP address for whatever reason idk, query it to get actual 100% correct one. It is similar as for retained mqtt messages, most of the time it generates more problems than it solves.

**Reply by joba-1** [reply]
> Corner case, why should DHCP server loose the associations and at the same time Tasmota device does a reconnect?

While I agree it is a corner case, just to answer your question:

because power outage and the dhcp server is e.g. a tasmota range expander that can not keep track of leases in flash?

**Reply by AlfaBravoX** [reply]
OK. i need to call off my findings. weird. it looks that problem lays at python library level. for some reason python PAHO lib is unable to read discovery messages from `tasmota/discovery/#` anymore, but mosquitto still client can. using mosquitto_sub I do see discovery messages re-appering after wifi reassociations after signal is lost. apologies for confusions I need to double check python stuff now

**Reply by barbudor** [reply]
You can always check what tasmota is doing
Discovery messages are displayed in the console with Weblog 3 or 4 (can't remember now but easy to check)

**Reply by sfromis** [reply]
Log level 4 (more debug) is needed to have those publish reflected in the Tasmota logs.
